enclave-runner/src/usercalls/mod.rs

```diff
@@ -76,148 +83,307 @@ impl<T> Write for ReadOnly<T> {
+                            continue
+                        }
+
+                        if tx.try_send(buf[..len].to_vec().into()).is_err() {
+                            return
+                        };
+                    }
+                });
+                Lock::new(AsyncStdin { rx, buf: VecDeque::new() })
+            };
+        }
+
+        let mut stdin = STDIN.clone().poll_lock().as_io_result()?;
+        if stdin.buf.is_empty() {
+            let pipeerr = || -> io::Error { io::ErrorKind::BrokenPipe.into() };
+            stdin.buf = stdin
+                .rx
+                .poll()
+                .map_err(|_| pipeerr())?
+                .as_io_result()?
+                .ok_or_else(pipeerr)?;
+        }
+        let inbuf = match stdin.buf.as_slices() {
+            (&[], inbuf) => inbuf,
+            (inbuf, _) => inbuf,
+        };
+        let len = cmp::min(buf.len(), inbuf.len());
+        buf[..len].copy_from_slice(&inbuf[..len]);
+        stdin.buf.drain(..len);
+        Ok(len)
     }
 }
 
-impl<'a> SharedStream<'a> for io::Stdin {
-    type Inner = ReadOnly<io::StdinLock<'a>>;
+pub trait AsyncStream: AsyncRead + AsyncWrite + 'static + Send + Sync {}
 
-    fn lock(&'a self) -> Self::Inner {
-        ReadOnly(io::Stdin::lock(self))
-    }
+impl<S: AsyncRead + AsyncWrite + Sync + Send + 'static> AsyncStream for S {}
+
+/// AsyncListener lets an implementation implement a slightly modified form of `std::net::TcpListener::accept`.
+pub trait AsyncListener: 'static + Send {
+    /// The enclave may optionally request the local or peer addresses
+    /// be returned in `local_addr` or `peer_addr`, respectively.
+    /// If `local_addr` and/or `peer_addr` are not `None`, they will point to an empty `String`.
+    /// On success, user-space can fill in the strings as appropriate.
+    ///
+    /// The enclave must not make any security decisions based on the local address received.
+    fn accept<'a>(
+        &'a mut self,
+        local_addr: Option<&'a mut String>,
+        peer_addr: Option<&'a mut String>,
+    ) -> std::pin::Pin<Box<dyn futures::Future<Output = IoResult<Box<dyn AsyncStream>>> + 'a>>;
 }
 
-impl<'a> SharedStream<'a> for io::Stdout {
-    type Inner = WriteOnly<io::StdoutLock<'a>>;
+struct AsyncStreamAdapter {
+    stream: Box<dyn AsyncStream>,
+    read_queue: VecDeque<tokio::prelude::task::Task>,
+    write_queue: VecDeque<tokio::prelude::task::Task>,
+    flush_queue: VecDeque<tokio::prelude::task::Task>,
+}
 
-    fn lock(&'a self) -> Self::Inner {
-        WriteOnly(io::Stdout::lock(self))
+fn notify_other_tasks(queue: &mut VecDeque<tokio::prelude::task::Task>) {
+    // NOTE: if you replace `is_current()` with `will_notify_current()`,
+    // you should also change the `if` to `while` in the next line because
+    // will_notify_current() is not required to return true.
+    queue.retain(|task| !task.will_notify_current());
+    while let Some(task) = queue.pop_front() {
```

Comment:
This is inefficient: first you delete some elements at random indices from the queue, requiring lots of memory shuffling. Then, you clear the queue. Why don't you just do this?
```rust
for task in queue.drain(..) {
    if !task.will_notify_current() {
        task.notify();
    }
}
```
Also I think the comment at the top of the function is obsolete.